src/data/set/intervals/instances.lean

```diff
@@ -0,0 +1,94 @@
+/-
+Copyright (c) 2022 Stuart Presnell. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Stuart Presnell, Eric Wieser, Yaël Dillies
+-/
+import data.set.intervals.basic
+import algebra.order.ring
+import algebra.group_power.order
+
+/-!
+# Structure instances for unit intervals
+
+For suitably structured underlying type `α`, we exhibit the structure of
+the unit intervals from `0` to `1`.
+
+## TODO
+* structure instances for intervals -1 to 1
+* provide `has_distrib_neg` instances where applicable
+-/
+
+variables {α : Type*} [ordered_semiring α]
+
+open set
+
+/-! ### Instances for `(Icc 0 1)` -/
+
+instance : has_zero (Icc (0:α) 1) := { zero := ⟨0, left_mem_Icc.2 zero_le_one⟩ }
+
+instance : has_one (Icc (0:α) 1) := { one := ⟨1, right_mem_Icc.2 zero_le_one⟩ }
+
+instance : has_mul (Icc (0:α) 1) :=
+{ mul := λ p q, ⟨p*q, ⟨mul_nonneg p.2.1 q.2.1, mul_le_one p.2.2 q.2.1 q.2.2⟩⟩ }
+
+instance : has_pow (Icc (0:α) 1) ℕ :=
+{ pow := λ p n, ⟨p.1 ^ n, ⟨pow_nonneg p.2.1 n, pow_le_one n p.2.1 p.2.2⟩⟩ }
+
+@[simp, norm_cast] lemma coe_Icc_zero : ↑(0 : Icc (0:α) 1) = (0 : α) := rfl
+@[simp, norm_cast] lemma coe_Icc_one : ↑(1 : Icc (0:α) 1) = (1 : α) := rfl
+@[simp, norm_cast] lemma coe_Icc_mul (x y : Icc (0:α) 1) : ↑(x * y) = (x * y : α) := rfl
+@[simp, norm_cast] lemma coe_Icc_pow (x : Icc (0:α) 1) (n : ℕ) : ↑(x ^ n) = (x ^ n : α) := rfl
+
+instance : monoid_with_zero (Icc (0:α) 1) :=
+by { apply function.injective.monoid_with_zero _ subtype.coe_injective; { intros, refl } }
+
+instance {α : Type*} [ordered_comm_semiring α] : comm_monoid_with_zero (Icc (0:α) 1) :=
```

Comment:
Either as a Todo or in this PR: `group_with_zero` on ordered fields

Reply:
Just checking: you mean `instance {α : Type*} [linear_ordered_field α] : group_with_zero (Icc (0:α) 1)`?  What's the inverse operation you have in mind?

Reply:
Oh, I'm being stupid!